src/materialized/bin/materialized.rs

```diff
@@ -288,3 +291,73 @@ to improve both our software and your queries! Please reach out at:
 "
     );
 }
+
+/// Attempts to increase the soft nofile rlimit to the maximum possible value.
+fn adjust_rlimits() {
+    // getrlimit/setrlimit can have surprisingly different behavior across
+    // platforms, even with the rlimit wrapper crate that we use. This function
+    // is chattier than normal at the trace log level in an attempt to ease
+    // debugging of such differences.
+
+    let (soft, hard) = match rlimit::Resource::NOFILE.get() {
+        Ok(limits) => limits,
+        Err(e) => {
+            trace!("unable to read initial nofile rlimit: {}", e);
+            return;
+        }
+    };
+    trace!("initial nofile rlimit: ({}, {})", soft, hard);
+
+    #[cfg(target_os = "macos")]
+    let hard = {
+        use std::cmp;
+        use sysctl::Sysctl;
+
+        // On macOS, getrlimit by default reports that the hard limit is
+        // unlimited, but there is usually a stricter hard limit discoverable
+        // via sysctl. Failing to discover this secret stricter hard limit will
+        // cause the call to setrlimit below to fail.
+        let res = sysctl::Ctl::new("kern.maxfilesperproc")
+            .and_then(|ctl| ctl.value())
+            .map_err(|e| e.to_string())
+            .and_then(|v| match v {
+                sysctl::CtlValue::Int(v) => Ok(v as u64),
+                o => Err(format!("unexpected sysctl value type: {:?}", o)),
+            });
+        match res {
+            Ok(v) => {
+                trace!("sysctl kern.maxfilesperproc hard limit: {}", v);
+                cmp::min(v, hard)
+            }
+            Err(e) => {
+                trace!("error while reading sysctl: {}", e);
+                hard
+            }
+        }
+    };
+
+    trace!("attempting to adjust nofile rlimit to ({0}, {0})", hard);
+    if let Err(e) = rlimit::Resource::NOFILE.set(hard, hard) {
```

Comment:
this looks like you're just setting the soft limit to the hard limit, instead of to `u16::MAX`, which is what I'd expect it to do?

Reply:
You can't set the soft limit to anything greater than the hard limit. That's the whole point of the hard limit :D. (The reason the API looks like this is because you might want to decrease the hard limit—but you are not permitted to increase it.) If the hard limit is unlimited, `hard` will be `u16::MAX`, and all will work out properly.